inst/include/dust/kernels.hpp

```diff
@@ -0,0 +1,123 @@
+#ifndef DUST_KERNELS_HPP
+#define DUST_KERNELS_HPP
+
+template <typename T>
+DEVICE void update_device(size_t step,
+                   const dust::interleaved<typename T::real_t> state,
+                   dust::interleaved<int> internal_int,
+                   dust::interleaved<typename T::real_t> internal_real,
+                   const int * shared_int,
+                   const typename T::real_t * shared_real,
+                   dust::rng_state_t<typename T::real_t>& rng_state,
+                   dust::interleaved<typename T::real_t> state_next);
+
+// __global__ for shuffling particles
+template<typename real_t>
+KERNEL void scatter_device(int* scatter_index,
+                           real_t* state,
+                           real_t* scatter_state,
+                           size_t state_size) {
+#ifdef __NVCC__
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i < state_size) {
```

Comment:
I have a new preference for these which makes them more similar to the non NVCC version below
```suggestion
  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < state_size; i += blockDim.x * gridDim.x) {
```
see: https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/